javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,229 @@
+   * Extending this class does nothing, its mostly to indicate intent.
+   *
+   * The classes EventRegistration::Range and EventDispatch::Range must be extended to get a working EventEmitter model.
+   * An EventRegistration models a method call that registers some event handler on an EventEmitter.
+   * And EventDispatch models that some event is dispatched on an EventEmitter.
+   *
+   * Both the EventRegistration and EventDispatch have a field `emitter`,
+   * which is the EventEmitter that events are registered on / dispatched to respectively.
+   *
+   * Here is a simple JavaScript example with the NodeJS EventEmitter:
+   *    var e = new EventEmitter(); // <- EventEmitter
+   *    e.on("name", (data) => {...}); // <- EventRegistration
+   *    e.emit("name", "foo"); // <- EventDispatch
+   */
+  abstract class Range extends DataFlow::Node { }
+}
+
+/**
+ * An EventEmitter instance that implements the EventEmitter API.
+ * Extend EventEmitter::Range to mark something as being an EventEmitter.
+ */
+class EventEmitter extends DataFlow::Node {
+  EventEmitter::Range range;
+
+  EventEmitter() { this = range }
+}
+
+/**
+ * A registration of an event handler on an EventEmitter.
+ */
+class EventRegistration extends DataFlow::Node {
+  EventRegistration::Range range;
+
+  EventRegistration() { this = range }
+
+  /** Gets the EventEmitter that the event handler is registered on. */
+  final EventEmitter getEmitter() { result = range.getEmitter() }
+
+  /** Gets the name of the channel if possible. */
+  string getChannel() { result = range.getChannel() }
+
+  /** Gets the `i`th parameter in the event handler. */
+  DataFlow::Node getReceivedItem(int i) { result = range.getReceivedItem(i) }
+
+  /**
+   * Gets a value that is returned by the event handler.
+   * The default implementation is that no value can be returned.
+   */
+  DataFlow::Node getAReturnedValue() { result = range.getAReturnedValue() }
+
+  /**
+   * Get a dispatch that this event handler can return a value to.
+   * The default implementation is that there exists no such dispatch.
+   */
+  EventDispatch getAReturnDispatch() { result = range.getAReturnDispatch() }
+}
+
+module EventRegistration {
+  /**
+   * A registration of an event handler on an EventEmitter.
+   */
+  abstract class Range extends DataFlow::Node {
+    EventEmitter::Range emitter;
+
+    final EventEmitter getEmitter() { result = emitter }
+
+    abstract string getChannel();
+
+    abstract DataFlow::Node getReceivedItem(int i);
+
+    abstract DataFlow::Node getAReturnedValue();
+
+    abstract EventDispatch::Range getAReturnDispatch();
+  }
+
+  /**
+   * A default implementation of an EventRegistration.
+   * The default implementation assumes that `this` is a DataFlow::InvokeNode where the
+   * first argument is a string describing which channel is registered, and the second
+   * argument is the event handler callback.
+   */
+  abstract class DefaultEventRegistration extends Range {
```

Comment:
How about adding the `InvokeNode` requirement to the charpred, and avoiding the explicit casts? This changes the semantics slightly, but I think it can avoid some confusion. 

```suggestion
  abstract class DefaultEventRegistration extends Range, DataFlow::InvokeNode {
```
This should also be done below in the `DefaultEventDispatch` class.